txnsync/msgbuffers.go

```diff
@@ -0,0 +1,76 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"sync"
+
+	"github.com/algorand/go-algorand/data/transactions"
+)
+
+const messageBufferDefaultInitialSize = 10240
+
+// msgBuffersPool holds temporary byte slice buffers used for encoding messages.
+var msgBuffersPool = sync.Pool{
+	New: func() interface{} {
+		return make([]byte, 0, messageBufferDefaultInitialSize)
+	},
+}
+
+// GetEncodingBuf returns a byte slice that can be used for encoding a
+// temporary message.  The byte slice has zero length but potentially
+// non-zero capacity.  The caller gets full ownership of the byte slice,
+// but is encouraged to return it using releaseMessageBuffer().
+func getMessageBuffer() []byte {
+	return msgBuffersPool.Get().([]byte)[:0]
+}
+
+// releaseMessageBuffer places a byte slice into the pool of temporary buffers
+// for encoding.  The caller gives up ownership of the byte slice when
+// passing it to releaseMessageBuffer().
+func releaseMessageBuffer(s []byte) {
+	msgBuffersPool.Put(s)
+}
+
+// txidSlicePool holds temporary byte slice buffers used for encoding messages.
+var txidSlicePool = sync.Pool{}
+
+// getTxIDSliceBuffer returns a slice that can be used for storing a
+// list of transaction IDs. The slice has zero length but potentially
+// non-zero capacity.  The caller gets full ownership of the slice,
+// but is encouraged to return it using releaseTxIDSliceBuffer().
+func getTxIDSliceBuffer(minSize int) []transactions.Txid {
```

Comment:
Golang docs for sync.Pool think this is a bad idea:

> a free list maintained as part of a short-lived object is not a suitable use for a Pool, since the overhead does not amortize well in that scenario

Reply:
The emphesis here was *short lived object*. The usage here is global, hence *long lived*...

Reply:
> I don't love this pattern and I think it's pretty weird. I assume you've done the benchmarking to show that this is better, but it comes with tradeoffs, right? This is memory that will never be returned to the OS if usage goes down. This is a memory pool that only grows. This is memory reserved for this one specific purpose. Should we have a maximum size of the pool? (e.g. 1000 idle buffers in the pool, after that they just get released to GC) Or possibly a maximum total bytes per pool? Buffer up to 100kB or 1000 instances of some type of thing?

The memory would be returned to the os within two GC calls. Memory pressure triggers GC, which would expedite that. The sync.Pool is quite optimized, allowing bypassing locks for thread local buffers.

This is valuable on relays, who have these buffers allocated frequently. I could write a benchmark for that - but it would be testing the exact reason sync.Pool was created for.

The one thing I'm not sure about is the threading model - I think that it's more than we need in some cases, and a custom implementation would outperform that. I'd suggest revisit this if we'll have CPU issues on that allocator.

Reply:
Note that in future PRs I think that we should "dress-up" this solution into a long-lived "memory pools" object, to avoid global variable.

Reply:
As for benchmark, I did do few memory benchmarks against the emulator to help me to identify the most-frequent memory allocators, and target them.